### PR TITLE
Broadcast address changes at Capella

### LINKED ISSUE #3901: Broadcast BLS to execution changes from op pool at Capella fork boundary
## Description

Related to the compromised address mitigations (see #3900), we should have Lighthouse broadcast BLS to execution changes at the Capella fork.

Today we (LH, Nimbus, Prysm) decided to adopt a LIFO packing strategy to discourage most users from spamming at the fork. Still, for users whose keys have been compromised, spamming is the best strategy (they need to get their messages to all honest nodes ASAP, particularly the ones that didn't opt-in to the agreed list).

The broadcast _also_ benefits ordinary users who want to get their address change in early by submitting it to their own op pool prior to the fork.

## Steps to resolve

One way to implement this might be a dedicated task that fires from `on_slot` as soon as Capella is hit. It could read from the pool to get a `Vec<Arc<SigVerifiedOp<..>>>`, and then broadcast them in batches. At the moment of broadcast it should verify that the message hasn't been seen on gossip yet, and is still valid (this may require fiddling with some of the other verification checks to get the order right).

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

Closes #3901

## Proposed Changes

Broadcasts `BlsToExecutionChanges` at the Capella fork. For an address change to be broadcast, it must have been received via the HTTP *before* the Capella fork.

Address changes will be broadcast in batches of 128 at 500ms intervals. This rate limiting helps protect the Ethereum network from a flood at messages at the start of the slot. It also helps ensure that our `network_send` channel isn't saturated with address changes messages, resulting in late broadcast of other messages (blocks, attestations, etc).

The address changes in the pool will be shuffled (`thread_rng`) before the batched broadcasting starts. This helps ensure fairness.

## Breaking Changes

This PR involves a database migration from v14 to v15. It's possible to downgrade using the `lighthouse db` tool.

### CODE REVIEW

File: beacon_node/http_api/src/lib.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]: Won't this always be `CapellaBroadcast::No` for messages received prior to the fork? Which will result in no changes being broadcast with a delay?

File: beacon_node/client/src/address_change_broadcast.rs
Comment (michaelsproul) [inline beacon_node/client/src/address_change_broadcast.rs:42]: Seeing as this function only runs once and _only_ if it lands at the fork exactly, do we need to store the broadcast indices on disk? I was thinking we could just snapshot the contents of the pool at the fork boundary, store that in-memory in this function and then call it a day once that snapshot is fully broadcast.

If the user restarts their node during this process they're going to miss out with either of these approaches. If we wanted to strengthen the guarantee to account for nodes that go offline we should allow broadcasting the pool contents at any point after Capella, if we have messages on-disk that are awaiting broadcast.

So depending on what we decide:

- Option 1: broadcast at any time after fork: need the indices on-disk.
- Option 2: broadcast only at the fork slot: no need to store indices on-disk, no need for a schema migration.

File: beacon_node/client/src/address_change_broadcast.rs
Comment (michaelsproul) [inline beacon_node/client/src/address_change_broadcast.rs:97]: Rather than publishing unconditionally I was wondering if we should check the gossip filter first. It's likely that particularly for later batches many of the messages will already have been propagated and seen.

The complication is that if a message is in our pool currently we will avoid updating our gossip filter at all when we see it on the network. We could probably fix this by reordering the conditions here:

https://github.com/sigp/lighthouse/blob/f615fb08851dad655f7aea3159da060a83622e44/beacon_node/beacon_chain/src/beacon_chain.rs#L2234-L2254

We'd also need a read-only message to check if a validator has been seen in the `observed_bls_to_execution_changes`.

**Review by michaelsproul — COMMENTED** [review]
Apologies for the quick drive-by review while this is still WIP.

I'd been thinking about the design for this feature and bit and wanted to check it out. I wrote some stuff up in this issue (#3901), but repeated here in context too.

File: beacon_node/client/src/address_change_broadcast.rs
Comment (paulhauner) [reply beacon_node/client/src/address_change_broadcast.rs:97]: > Rather than publishing unconditionally I was wondering if we should check the gossip filter first. It's likely that particularly for later batches many of the messages will already have been propagated and seen.

My understanding is that the gossip network will natively de-dupe messages it's already seen. I'm yet to confirm this, but that's what I was working from.

Comment (paulhauner) [reply beacon_node/client/src/address_change_broadcast.rs:97]: > I'm yet to confirm this

I just confirmed this, we won't republish on the network.

File: beacon_node/client/src/address_change_broadcast.rs
Comment (paulhauner) [reply beacon_node/client/src/address_change_broadcast.rs:42]: > Seeing as this function only runs once and only if it lands at the fork exactly

This is all still WIP, so I wouldn't put too much weight on the peculiarities of the timing here. I'll give it all another run-through and try and get it to be sane.

> we should allow broadcasting the pool contents at any point after Capella

This is something @potuz is working on over in https://github.com/prysmaticlabs/prysm/pull/11871. I think it's an useful idea and I've been holding off on adding this functionality until that PR is merged. I figure we can try and match the URL.

Comment (potuz) [reply beacon_node/client/src/address_change_broadcast.rs:42]: Now that I've been tagged, the reason we haven't merged it (or that I haven't pushed back the resistance to have custom endpoints) is that a user can simply restart and resubmit his changes to the endpoint. I do think there's some minimal benefit in allowing a user to resubmit everything including all messages that were obtained in gossip so that less messages are lost by dropping

Comment (paulhauner) [reply beacon_node/client/src/address_change_broadcast.rs:42]: Alrighty, I've implemented the following:

- Indices requiring Capella broadcast are stored on disk.
- The broadcast routine runs both at the Capella fork and each time the node starts.
- After each address change is broadcast, the relevant val index is removed from the list.
- There is no way to re-trigger the broadcast of address changes.

This provides *some* guarantee that any pre-Capella address changes will eventually be sent to our networking service, regardless of reboots. I'm sure there's some edge-cases where we don't manage to actually get a message onto the network, however I don't think we need to cover these scenarios.

I've avoided the broadcast re-triggering logic since I think it's scope creep on this PR. We can add it later if we want it.

File: beacon_node/client/src/address_change_broadcast.rs
Comment (michaelsproul) [inline beacon_node/client/src/address_change_broadcast.rs]:
```suggestion
#[cfg(not(debug_assertions))] // Tests run too slow in debug.
```

Comment (paulhauner) [reply beacon_node/client/src/address_change_broadcast.rs]: Damn, thanks!

File: beacon_node/operation_pool/src/lib.rs
Comment (michaelsproul) [inline beacon_node/operation_pool/src/lib.rs]:
```suggestion
    pub fn get_bls_to_execution_changes_received_pre_capella(
```

**Review by michaelsproul — APPROVED** [review]
Looks great modulo one minor nit!

The tests are extra nice

File: beacon_node/operation_pool/src/lib.rs
Comment (paulhauner) [reply beacon_node/operation_pool/src/lib.rs]: Ah yep, good catch. This was a fumbled find-and-replace!

Comment (paulhauner) [reply beacon_node/operation_pool/src/lib.rs]: Fixed in https://github.com/sigp/lighthouse/pull/3919/commits/f45a1c87db58d0de9d38a09f6ccfe712c73a71d3

**Comment by michaelsproul** [comment]
Happy to squerge!

**Pull request merged**
